src/kernel_density_estimation.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+/* @flow */
+
+var stddev = require('./sample_standard_deviation');
+var interquartileRange = require('./interquartile_range');
+
+var SQRT_2PI = Math.sqrt(2 * Math.PI);
+
+/**
+ * [Well-known kernels](https://en.wikipedia.org/wiki/Kernel_(statistics)#Kernel_functions_in_common_use)
+ * @private
+ */
+var kernels /*: {[string]: (number) => number} */ = {
+    /**
+     * The gaussian kernel.
+     * @private
+     */
+    gaussian: function (u) {
+        return Math.exp(-0.5 * u * u) / SQRT_2PI;
+    }
+};
+
+/**
+ * Well known bandwidth selection methods
+ * @private
+ */
+var bandwidthMethods /*: {[string]: (Array<number>) => number} */ = {
+    /**
+     * The ["normal reference distribution"
+     * rule-of-thumb](https://stat.ethz.ch/R-manual/R-devel/library/MASS/html/bandwidth.nrd.html),
+     * a commonly used version of [Silverman's
+     * rule-of-thumb](https://en.wikipedia.org/wiki/Kernel_density_estimation#A_rule-of-thumb_bandwidth_estimator).
+     * @private
+     */
+    nrd: function (x /*: Array<number> */) {
+        var s = stddev(x);
+        var iqr = interquartileRange(x);
+        if (typeof iqr === 'number') {
+            s = Math.min(s, iqr / 1.34)
+        }
+        return 1.06 * s * Math.pow(x.length, -0.2);
+    }
+}
+
+/**
+ * [Kernel density estimation](https://en.wikipedia.org/wiki/Kernel_density_estimation)
+ * is a useful tool for, among other things, estimating the shape of the
+ * underlying probability distribution from a sample.
+ *
+ * @param X sample values
+ * @param kernel The kernel function to use. If a function is provided, it should return non-negative values and integrate to 1. Defaults to 'gaussian'.
+ * @param bandwidthMethod The "bandwidth selection" method to use, or a fixed bandwidth value. Defaults to "nrd", the commonly-used ["normal reference distribution" rule-of-thumb](https://stat.ethz.ch/R-manual/R-devel/library/MASS/html/bandwidth.nrd.html).
```

Comment:
Duplicated the description/link for normal reference distribution here since, as of now, the `bandwidthMethods` object is private.